### DISCUSSION TITLE
Markdown Image URLs

### DISCUSSION THREAD

**Discussion by digitaljohn** [directus/directus#5529] [Feature Requests]
We are aiming for all assets/files managed by Directus to be stored in storage buckets and also have the frontend application load these files directly rather than proxy through Directus.

We have a workable solution for non-markdown/general image fields but are blocked with images contained within markdown. Currently markdown content comes through with images with links to the absolute path to the image proxied through Directus.

Would it be possible to have a configurable option when setting up the field that controls this behaviour? So maybe the "Image Token" functionality could be renamed or replaced?

E.g. Make it default to something like `{directus_image_url}` so it replicates the existing default behaviour and still allows you to change this to things like `{directus_image_url}?format=webp` as intended. And in our use-case, we would set it to just `{filename_disk}` allowing us to prepend the bucket details on the frontend application.

**Comment by digitaljohn** [comment]
Looking at the source code I feel my initial understanding of the existing "Image Token" functionality was incorrect.

Regardless, there appears to be no way of us analysing the image URL output into markdown to extrapolate the location on the storage bucket. Do you feel this is outside the 80/20 rule and 80% of users will be happy with serving images proxied through Directus?

I do not mind working on this and pushing a PR if you think it's a worthwhile addition.

**Reply by rijkvanzanten** [reply]
We recommend loading the images through Directus, as that's how permissions are utilized, how thumbnails can be generated, etc. It's not a given that any configured storage adapter even has a public URL (eg S3 bucket could have it's files set to private, local folder might not be exposed to the internet, etc), so there's no way for Directus to have access to those paths / urls in the first place 👍🏻 

I don't think this fits the 80/20 for the markdown interface. This configuration would open up a way bigger can of worms than you'd might think. I'd recommend either using a custom hook to process the image paths on create/update, or create a custom md interface that saves it with your specific needs 👍🏻

**Reply by digitaljohn** [reply]
Interesting... it looks pretty simple to just copy your current markdown interface and tweak this function:
https://github.com/directus/directus/blob/acd41eb0be90933db09af6b02d3ffe0c56a64efd/app/src/interfaces/markdown/markdown.vue#L291-L303

Just confirming there is no way to granularly override existing behaviour?

**Reply by rijkvanzanten** [reply]
No, there's no option to override that 👍🏻

**Reply by rokdd** [reply]
I am also struggeling with the public url in the image paths - because we have different stages, but the path of the images is always targeting the initial one. I also asked myself what happens when I change the public url after uploading? Is there any update around this topic?

**Reply by paescuj** [reply]
One idea would be to replace the public URL by an custom hook / flow, see https://github.com/directus/directus/discussions/7700.

**Comment by luciandavidescu** [comment]
Can be done easily in a 3-block flow. On upload, this would make either a copy or a symlink of the uuid file, with the "download" name. Start with an action event hook on _files.upload_. Then in a _"Run script"_ block do something like this:

```
// fs-extra comes with directus, but you do need to enable it in the .env file
const fs = require('fs-extra')
module.exports = async function(data) {
    let originals = '/directus/uploads/';
    let copies = '/public/folder/';
    let input = {{$trigger}}.payload.filename_disk;
    let output = {{$trigger}}.payload.filename_download;

//this is to avoid name collisions, it will check for existence and increment by one
    let base = output.split('.').slice(0,-1).join('.');
    let type = output.split('.').pop();
        let counter = 0;
        while (fs.existsSync(copies+output)) {
            counter++;output = base + '-' + counter + '.' + type;
        }

// either a symlink like this or make a copy
        fs.symlink(originals+input,copies+output,'file',(err) => {})

// you'll need to write this back into the db (directus_files/id/filename_download) after, in case it was incremented
    return {output}
}
```

Last do an "update data" on directus_files with ID {{$trigger.key}} and the payload `{ "filename_download": "{{$last.output}}"}`.

Ensuring consistency of "file_download" will help achieve the reverse operation and remove the copy when the original is deleted.

To delete:
1. Do another flow, for on directus_files items.delete, this time a filter.
2. Get the download filenames from the db with a _"read data"_ on _directus_files_ with id `{{$trigger.payload}}` and optionally the query `{"fields": "filename_download"}`.
3. Do a "run script" block like this:
    
```
    const fs = require('fs-extra');
module.exports = async function(data) {
    let files = [].concat({{$last}});
    for (let file of files) {
        fs.unlink('/public/folder/' + file.filename_download, (err) => {});
    }
}
```

**Reply by luciandavidescu** [reply]
(@rokdd not sure how notifications work so I hope you see this)

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
